SimpleMockNioTransportTests.testConnectException fails on MacOS

This seems to be triggered by #32560 (reverting 0cf0d73813e7732234e2a89611f05b9faa165ad2 solves the problem)

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-darwin-compatibility/11/console

```
REPRODUCE WITH: ./gradlew :test:framework:test -Dtests.seed=9C3DC2779BDE41C3 -Dtests.class=org.elasticsearch.transport.nio.SimpleMockNioTransportTests -Dtests.method="testConnectException" -Dtests.security.manager=true -Dtests.locale=ar-LB -Dtests.timezone=America/Anchorage -Dcompiler.java=10 -Druntime.java=8
```

```
02:08:43   1> [2018-09-19T18:08:40,751][INFO ][o.e.t.t.MockTransportService] [testConnectException] publish_address {127.0.0.1:49748}, bound_addresses {[::1]:49747}, {127.0.0.1:49748}
02:08:43   1> [2018-09-19T18:08:40,773][INFO ][o.e.t.n.SimpleMockNioTransportTests] [testConnectException] after test
02:08:43 FAILURE 0.06s J3 | SimpleMockNioTransportTests.testConnectException <<< FAILURES!
02:08:43    > Throwable #1: java.lang.AssertionError: 
02:08:43    > Expected: a string containing "connect_exception"
02:08:43    >      but: was "[][127.0.0.1:9876] general node connection failure"
02:08:43    > 	at __randomizedtesting.SeedInfo.seed([9C3DC2779BDE41C3:78B03D75C4C593DD]:0)
02:08:43    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
02:08:43    > 	at org.elasticsearch.transport.nio.SimpleMockNioTransportTests.testConnectException(SimpleMockNioTransportTests.java:107)
02:08:43    > 	at java.lang.Thread.run(Thread.java:748)
02:08:43   1> [2018-09-19T18:08:40,793][INFO ][o.e.t.n.SimpleMockNioTransportTests] [testErrorMessage] before test
```

Reproduces for me `Darwin Kernel Version 16.7.0: Fri Apr 27 17:59:46 PDT 2018; root:xnu-3789.73.13~1/RELEASE_X86_64`

CC: @tbrooks8

--- a/test/framework/src/main/java/org/elasticsearch/transport/nio/MockNioTransport.java
+++ b/test/framework/src/main/java/org/elasticsearch/transport/nio/MockNioTransport.java
@@ -277,8 +277,9 @@ public class MockNioTransport extends TcpTransport {
 
         @Override
         public void setSoLinger(int value) throws IOException {
-            if (isOpen()) {
-                getRawChannel().setOption(StandardSocketOptions.SO_LINGER, value);
+            SocketChannel rawChannel = getRawChannel();
+            if (rawChannel.isConnected()) {
+                rawChannel.setOption(StandardSocketOptions.SO_LINGER, value);
             }
         }
 

--- a/test/framework/src/test/java/org/elasticsearch/transport/nio/SimpleMockNioTransportTests.java
+++ b/test/framework/src/test/java/org/elasticsearch/transport/nio/SimpleMockNioTransportTests.java
@@ -19,7 +19,6 @@
 
 package org.elasticsearch.transport.nio;
 
-import org.apache.lucene.util.Constants;
 import org.elasticsearch.Version;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
@@ -100,7 +99,6 @@ public class SimpleMockNioTransportTests extends AbstractSimpleTransportTestCase
     }
 
     public void testConnectException() throws UnknownHostException {
-        assumeFalse("Broken on Darwin - https://github.com/elastic/elasticsearch/issues/33879", Constants.MAC_OS_X);
         try {
             serviceA.connectToNode(new DiscoveryNode("C", new TransportAddress(InetAddress.getByName("localhost"), 9876),
                 emptyMap(), emptySet(),Version.CURRENT));

